### DISCUSSION TITLE
Error with For loop and matrix multiplication in PYTHON

### DISCUSSION THREAD

**Discussion by ramkarumugam** [community/community#21389] [Other Feature Feedback, Questions, & Ideas]
<pre><code>ValueErrorTraceback (most recent call last)
\&lt;ipython-input-35-1ecd9c5afb6f\&gt;in\&lt;module\&gt; 19print(right_temp.size) 20print(coordinates_left,coordinates_right)---\&gt; 21same,h,w=if_same(left_temp,right_temp) 22#plt.figure(figsize=[h,w]) 23#plt.imshow(out1)\&lt;ipython-input-33-27346e0a4cef\&gt;inif\_same(left\_temp, right\_temp) 32image1=left_temp_np[0:h][0:w] 33image2=right_temp_np[j:h+j][i:w+i]---\&gt; 34same=mse(image1,image2) 35print("loop 1 i: %s j: %s image1: %s image2: %s"%(i,j,image1.size,image2.size)) 36print(same)\&lt;ipython-input-34-3ff65825a155\&gt;inmse(image1, image2) 1defmse(image1,image2):----\&gt; 2diff=(image1.astype(float)-image2.astype(float)) 3error=diff\*\*2 4[w,h]=image1.shape
 5den=w\*h

ValueError: operands could not be broadcast together with shapes (10,10) (10,11)
</code></pre>
<p>Above is the Error. </p>
<p>The code employs if conditions and for loop to scans unequal images from two cameras and checks for matches. </p>
<p><strong>’if_same’</strong> : This is the main function which scans and checks. </p>
<p>‘mse’ calls another function which calculates squared error.</p>
<pre><code>def mse(image1, image2):
    diff=(image1.astype(float)-image2.astype(float))
    error=diff**2
    [w,h]=image1.shape
    den=w*h
    mse=error/den

    return mse

def if_same(left_temp, right_temp):

    h_L, w_L = left_temp.size
    h_R, w_R = right_temp.size

    left_temp_np=np.asarray(left_temp)
    right_temp_np=np.asarray(right_temp)

    diff_w=abs(w_L-w_R)+1
    diff_h=abs(h_L-h_R)+1

    image1=[]
    image2=[]
    #same= np.zeros([h,w], np.uint8)
    same=0
    h=0
    w=0

    if w_L&lt;=w_R:
        if h_L&lt;=h_R:
            [w,h]=[w_L,h_L]
            #image1=np.zeros[h,w]
            #image2=np.zeros[h,w]
            i=0
            for i in range(0,diff_w):
                j=0
                for j in range(0,diff_h):
                    print(j, h, h+j, i, w, w+i)
                    image1=left_temp_np[0:h][0:w]
                    image2=right_temp_np[j:h+j][i:w+i]
                    same=mse(image1, image2)
                    print("loop 1 i: %s j: %s image1: %s image2: %s" % (i,j,image1.size,image2.size))
                    print(same)

        elif h_R&lt;h_L:
            [w,h]=[w_L,h_R]
            i=0
            for i in range(0,diff_w):
                j=0
                for j in range(0,diff_h):
                    image1=left_temp_np[j:h+j][0:w]
                    image2=right_temp_np[0:h][i:w+i]
                    same=mse(image1, image2)
                    print("loop 2 i: %s j: %s image1: %s image2: %s" % (i,j,image1.size,image2.size))
                    print(same)
        return same, h, w

    elif w_R&lt;w_L:
        if h_L&lt;=h_R:
            i=0
            [w,h]=[w_R,h_L]
            for i in range(0,diff_w):
                j=0
                for j in range(0,diff_h):
                    image1=left_temp_np[0:h][i:w+i]
                    image2=right_temp_np[j:h+j][0:w]
                    same=mse(image1, image2)
                    print("loop 3 i: %s j: %s image1: %s image2: %s" % (i,j,image1.size,image2.size))
                    print(same)

        elif h_R&lt;h_L:
            i=0
            [w,h]=[w_R,h_R]
            for i in range(0,diff_w):
                j=0
                for j in range(0,diff_h):
                    print(j)
                    image1=left_temp_np[j:(h+j)][i:(w+i)]
                    image2=right_temp_np[0:h][0:w]
                    print("loop 4 i: %s j: %s image1: %s image2: %s" % (i,j,image1.size,image2.size))
                    same=mse(image1, image2)
                    print(same)
        return same, h, w

    return same, h, w
</code></pre>
<p>I have actually tried printing the i,j,w,h  to see if there is the problem. </p>
<pre><code>0 10 10 0 13 13
</code></pre>
<p>The code should slice matrixes of dimensions 10x13 from both images. But I get 10x10 and 10x11. </p>
<p>If the slice is [0:10][0:13] and [1:11][1:14], both will return [10x13]. Correct me, if I’m wrong</p>

**Comment by ramkarumugam** [comment, marked as answer]
<p>The shape of the image gets inversed when read as an array. </p>
<p>I later found that out by printing the shapes of images and image arrays. </p>
<blockquote>
<p><a class="mention" href="/ramkarumugam">@ramkarumugam</a> wrote:</p>
</blockquote>
<blockquote>
<p>ValueErrorTraceback (most recent call last)<br>
&lt;ipython-input-35-1ecd9c5afb6f&gt;in&lt;module&gt; 19print(right_temp.size) 20print(coordinates_left,coordinates_right)—&gt; 21same,h,w=if_same(left_temp,right_temp) 22#plt.figure(figsize=[h,w]) 23#plt.imshow(out1)&lt;ipython-input-33-27346e0a4cef&gt;inif_same(left_temp, right_temp) 32image1=left_temp_np[0:h][0:w] 33image2=right_temp_np[j:h+j][i:w+i]—&gt; 34same=mse(image1,image2) 35print(“loop 1 i: %s j: %s image1: %s image2: %s”%(i,j,image1.size,image2.size)) 36print(same)&lt;ipython-input-34-3ff65825a155&gt;inmse(image1, image2) 1defmse(image1,image2):----&gt; 2diff=(image1.astype(float)-image2.astype(float)) 3error=diff**2 4[w,h]=image1.shape<br>
5den=w*h</p>
</blockquote>
<blockquote>
<p>ValueError: operands could not be broadcast together with shapes (10,10) (10,11)</p>
</blockquote>
<blockquote>
<p>Above is the Error. </p>
</blockquote>
<blockquote>
<p> </p>
</blockquote>
<blockquote>
<p>The code employs if conditions and for loop to scans unequal images from two cameras and checks for matches. </p>
</blockquote>
<blockquote>
<p> </p>
</blockquote>
<blockquote>
<p><strong>’if_same’</strong> : This is the main function which scans and checks. </p>
</blockquote>
<blockquote>
<p>‘mse’ calls another function which calculates squared error.</p>
</blockquote>
<blockquote>
<p>def mse(image1, image2):<br>
diff=(image1.astype(float)-image2.astype(float))<br>
error=diff**2<br>
[w,h]=image1.shape<br>
den=w*h<br>
mse=error/den</p>
</blockquote>
<blockquote>
<p>return mse</p>
</blockquote>
<blockquote>
<p> </p>
</blockquote>
<blockquote>
<p> </p>
</blockquote>
<blockquote>
<p>def if_same(left_temp, right_temp):</p>
</blockquote>
<blockquote>
<p>h_L, w_L = left_temp.size<br>
h_R, w_R = right_temp.size</p>
</blockquote>
<blockquote>
<p>left_temp_np=np.asarray(left_temp)<br>
right_temp_np=np.asarray(right_temp)</p>
</blockquote>
<blockquote>
<p>diff_w=abs(w_L-w_R)+1<br>
diff_h=abs(h_L-h_R)+1</p>
</blockquote>
<blockquote>
<p>image1=[]<br>
image2=[]<br>
#same= np.zeros([h,w], np.uint8)<br>
same=0<br>
h=0<br>
w=0</p>
</blockquote>
<blockquote>
<p>if w_L&lt;=w_R:<br>
if h_L&lt;=h_R:<br>
[w,h]=[w_L,h_L]<br>
#image1=np.zeros[h,w]<br>
#image2=np.zeros[h,w]<br>
i=0<br>
for i in range(0,diff_w):<br>
j=0<br>
for j in range(0,diff_h):<br>
print(j, h, h+j, i, w, w+i)<br>
image1=left_temp_np[0:h][0:w]<br>
image2=right_temp_np[j:h+j][i:w+i]<br>
same=mse(image1, image2)<br>
print(“loop 1 i: %s j: %s image1: %s image2: %s” % (i,j,image1.size,image2.size))<br>
print(same)</p>
</blockquote>
<blockquote>
<p>elif h_R&lt;h_L:<br>
[w,h]=[w_L,h_R]<br>
i=0<br>
for i in range(0,diff_w):<br>
j=0<br>
for j in range(0,diff_h):<br>
image1=left_temp_np[j:h+j][0:w]<br>
image2=right_temp_np[0:h][i:w+i]<br>
same=mse(image1, image2)<br>
print(“loop 2 i: %s j: %s image1: %s image2: %s” % (i,j,image1.size,image2.size))<br>
print(same)<br>
return same, h, w</p>
</blockquote>
<blockquote>
<p>elif w_R&lt;w_L:<br>
if h_L&lt;=h_R:<br>
i=0<br>
[w,h]=[w_R,h_L]<br>
for i in range(0,diff_w):<br>
j=0<br>
for j in range(0,diff_h):<br>
image1=left_temp_np[0:h][i:w+i]<br>
image2=right_temp_np[j:h+j][0:w]<br>
same=mse(image1, image2)<br>
print(“loop 3 i: %s j: %s image1: %s image2: %s” % (i,j,image1.size,image2.size))<br>
print(same)</p>
</blockquote>
<blockquote>
<p>elif h_R&lt;h_L:<br>
i=0<br>
[w,h]=[w_R,h_R]<br>
for i in range(0,diff_w):<br>
j=0<br>
for j in range(0,diff_h):<br>
print(j)<br>
image1=left_temp_np[j:(h+j)][i:(w+i)]<br>
image2=right_temp_np[0:h][0:w]<br>
print(“loop 4 i: %s j: %s image1: %s image2: %s” % (i,j,image1.size,image2.size))<br>
same=mse(image1, image2)<br>
print(same)<br>
return same, h, w</p>
</blockquote>
<blockquote>
<p>return same, h, w</p>
</blockquote>
<blockquote>
<p>I have actually tried printing the i,j,w,h  to see if there is the problem. </p>
</blockquote>
<blockquote>
<p>0 10 10 0 13 13</p>
</blockquote>
<blockquote>
<p>The code should slice matrixes of dimensions 10x13 from both images. But I get 10x10 and 10x11. </p>
</blockquote>
<blockquote>
<p> </p>
</blockquote>
<blockquote>
<p>If the slice is [0:10][0:13] and [1:11][1:14], both will return [10x13]. Correct me, if I’m wrong</p>
</blockquote>
<blockquote>
<p> </p>
</blockquote>

**Comment by mpboom** [comment]
<p>Thanks so much for sharing your solution with the community. Is your problem fully resolved now?</p>

**Comment by ramkarumugam** [comment]
<p>Yeah.It’s solved fully. Thanks for responding Mark. </p>
<p>–Ram K</p>
